kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/AddFile.java

```diff
@@ -93,4 +103,208 @@ public static Row convertDataFileStatus(
+    this.size = size;
+    this.modificationTime = modificationTime;
+    this.dataChange = dataChange;
+    this.deletionVector = deletionVector;
+    this.tags = tags;
+    this.baseRowId = baseRowId;
+    this.defaultRowCommitVersion = defaultRowCommitVersion;
+    this.stats = stats;
+  }
+
+  public Row toRow() {
+    Map<Integer, Object> valueMap = new HashMap<>();
+    valueMap.put(COL_NAME_TO_ORDINAL.get("path"), path);
+    valueMap.put(COL_NAME_TO_ORDINAL.get("partitionValues"), partitionValues);
+    valueMap.put(COL_NAME_TO_ORDINAL.get("size"), size);
+    valueMap.put(COL_NAME_TO_ORDINAL.get("modificationTime"), modificationTime);
+    valueMap.put(COL_NAME_TO_ORDINAL.get("dataChange"), dataChange);
+    deletionVector.ifPresent(dv -> valueMap.put(COL_NAME_TO_ORDINAL.get("deletionVector"), dv));
+    tags.ifPresent(tags -> valueMap.put(COL_NAME_TO_ORDINAL.get("tags"), tags));
+    baseRowId.ifPresent(rowId -> valueMap.put(COL_NAME_TO_ORDINAL.get("baseRowId"), rowId));
+    defaultRowCommitVersion.ifPresent(
+        commitVersion ->
+            valueMap.put(COL_NAME_TO_ORDINAL.get("defaultRowCommitVersion"), commitVersion));
+    stats.ifPresent(
+        stats -> valueMap.put(COL_NAME_TO_ORDINAL.get("stats"), stats.serializeAsJson()));
+    return new GenericRow(FULL_SCHEMA, valueMap);
+  }
+
+  public Optional<Long> getBaseRowId() {
+    return baseRowId;
+  }
+
+  public Optional<Long> getDefaultRowCommitVersion() {
+    return defaultRowCommitVersion;
+  }
+
+  public Optional<DataFileStatistics> getStats() {
+    return stats;
+  }
+
+  public Optional<Long> getNumRecords() {
+    return stats.map(DataFileStatistics::getNumRecords);
+  }
+
+  /** Creates a new AddFile instance with the specified base row ID. */
+  public AddFile withNewBaseRowId(long baseRowId) {
+    return new AddFile(
+        path,
+        partitionValues,
+        size,
+        modificationTime,
+        dataChange,
+        deletionVector,
+        tags,
+        Optional.of(baseRowId),
+        defaultRowCommitVersion,
+        stats);
+  }
+
+  /** Creates a new AddFile instance with the specified default row commit version. */
+  public AddFile withNewDefaultRowCommitVersion(long defaultRowCommitVersion) {
+    return new AddFile(
+        path,
+        partitionValues,
+        size,
+        modificationTime,
+        dataChange,
+        deletionVector,
+        tags,
+        baseRowId,
+        Optional.of(defaultRowCommitVersion),
+        stats);
+  }
+
+  @Override
+  public String toString() {
+    // Explicitly convert the partitionValues and tags to Java Maps
+    Map<String, String> partitionValuesJavaMap = VectorUtils.toJavaMap(this.partitionValues);
+    Optional<Map<String, String>> tagsJavaMap = this.tags.map(VectorUtils::toJavaMap);
+
+    StringBuilder sb = new StringBuilder();
+    sb.append("AddFile{");
+    sb.append("path='").append(path).append('\'');
+    sb.append(", partitionValues=").append(partitionValuesJavaMap);
```

Comment:
I think that partitionValues would be better printed out as the in-order `part1=value1, part2=value2` string

this map has no order guarantees, it could be confusing if we printed out `{ part2=value2, part1=value1 }`

The updated to AddFile alone, btw, would warrant their own PR. You could add an `AddFileSuite` to test these methods

Reply:
> I think that partitionValues would be better printed out as the in-order part1=value1, part2=value2 string

Yes, I’ll use a TreeMap to ensure the order.

> The updated to AddFile alone, btw, would warrant their own PR. You could add an AddFileSuite to test these methods

My initial intention was to combine them to better showcase how `AddFile` is used in row tracking and to justify the changes. I agree the updates to `AddFile` alone do need their own PR. I’ll have a separate PR for it with an `AddFileSuite` for testing.👍

Reply:
> I think that partitionValues would be better printed out as the in-order `part1=value1, part2=value2` string

Yes, I’ll use a `TreeMap` to ensure the order.

> The updated to AddFile alone, btw, would warrant their own PR. You could add an `AddFileSuite` to test these methods

I agree the updates to `AddFile` alone do need their own PR. My initial intention was to combine them to better showcase how `AddFile` is used in row tracking and to justify the changes. I can prepare a separate PR for it.